crates/uv/src/commands/pip/operations.rs

```diff
@@ -381,28 +383,106 @@ pub(crate) enum Modifications {
     Exact,
 }
 
+/// A distribution which was changed
+#[derive(Debug, Clone, Eq)]
+#[allow(clippy::large_enum_variant)]
+pub(crate) enum ChangedDist {
+    Local(LocalDist),
+    Remote(Arc<Dist>),
+}
+
+impl Hash for ChangedDist {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name().hash(state);
+        // TODO(tk): Use canonical
+        self.installed_version().hash(state);
+    }
+}
+
+impl PartialEq for ChangedDist {
+    fn eq(&self, other: &Self) -> bool {
+        // TODO(tk): Use canonical
+        self.name() == other.name() && self.installed_version() == other.installed_version()
+    }
+}
+
+impl From<LocalDist> for ChangedDist {
+    fn from(dist: LocalDist) -> Self {
+        Self::Local(dist)
+    }
+}
+
+impl From<InstalledDist> for ChangedDist {
+    fn from(dist: InstalledDist) -> Self {
+        Self::from(LocalDist::from(dist))
+    }
+}
+
+impl From<CachedDist> for ChangedDist {
+    fn from(dist: CachedDist) -> Self {
+        Self::from(LocalDist::from(dist))
+    }
+}
+
+impl From<Arc<Dist>> for ChangedDist {
+    fn from(dist: Arc<Dist>) -> Self {
+        Self::Remote(dist)
+    }
+}
+
+impl Name for ChangedDist {
+    fn name(&self) -> &PackageName {
+        match self {
+            Self::Local(dist) => dist.name(),
+            Self::Remote(dist) => dist.name(),
+        }
+    }
+}
+
+/// Zero version number
+static ZERO_VERSION: LazyLock<uv_pep440::Version> = LazyLock::new(|| uv_pep440::Version::new([0]));
```

Comment:
This is awful, it's a direct side effect of `ChangeEvent` requiring this type implements `InstalledMetadata`. I think that potentially, since `ChangeEvent` is only used in the logger and in uv-sync, it may make sense to use a different type and then this whole thing can be avoided.